ccl/sqlccl: add a load csv command

madelynnblue opened the pull request:
This adds the ability to convert a CSV file to enterprise BACKUP
format using only the CLI (without connecting to any server). This
is a good-enough solution for most of our current needs. A more
complicated, distributed CSV processing and sorting algorithm could
be used in the future for very large (> many hundred GB) files.

It works by creating a pipeline of channels with worker go routines
at each stage. The stages are: read the CSV into records per line,
convert the records into KV pairs, sort all KV pairs and split into
SST files, write SST files and create the backup descriptor. Only
the second stage (conversion) can be parallelized with multpile go
routines because all others require reading or writing in order.

Although we would like CSV importing (including the RESTORE step)
to not require an enterprise license, for now this code is CCL. When
we figure out exactly how we want to do that we may move it to Apache.

Closes #15933

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16611)
<!-- Reviewable:end -->

danhhz commented:
Haven't gotten to the tests yet, but did an initial pass on everything else. This is looking great! I really like the structure of the code; there's a lot going on here and I thought it was going to be a lot harder to review

---

Review status: 0 of 7 files reviewed at latest revision, 32 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn55h-UwFJbiupy7Nz-:-Kn55h-UwFJbiupy7Nz0:bk4qkqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L43)):*
> ```Go
> 
> func init() {
> 	loadCSVCmd := &cobra.Command{
> ```

This should live in a new cliccl package

---

*[pkg/ccl/sqlccl/csv.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn55s4bt-5sV38pyTAf:-Kn55s4bt-5sV38pyTAg:b-jk8naa) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L44)):*
> ```Go
> func init() {
> 	loadCSVCmd := &cobra.Command{
> 		Use:   "csv table=name dest=directory [data=file [...]] [comment=comment] [delimiter=delim] [nullif=null] [overwrite=true]",
> ```

I'd prefer these to all have "--" or "-" prefixes so they look more like flags. Here and below

---

*[pkg/ccl/sqlccl/csv.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn56sAWnxFfvjXZPm4k:-Kn56sAX4ygGyDnMrJZn:bsvx2ph) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L53)):*
> ```Go
> comment are ignored. Fields are considered null if equal to nullif
> (may be the empty string). Created files by default do not overwrite
> existing files unless overwrite is true.
> ```

Let's remove this overwrite option. It introduces tons of complexity below and doesn't add much value

---

*[pkg/ccl/sqlccl/csv.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn575VrzyJRo_Li-3I2:-Kn575VrzyJRo_Li-3I3:b-ailkuf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L56)):*
> ```Go
> 
> It requires approximately 2x the size of the data files of free disk
> space. An intermediate copy will be stored in the temp directory,
> ```

s/temp/os temp/

---

*[pkg/ccl/sqlccl/csv.go, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn57O2ruGeVXI6AEVeD:-Kn57O2ruGeVXI6AEVeE:b-cd87jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L78)):*
> ```Go
> 		Use:   "load [command]",
> 		Short: "loading commands",
> 		Long:  `Various commands for loading external files.`,
> ```

s/loading/bulk loading/ (also you can remove "Various")

---

*[pkg/ccl/sqlccl/csv.go, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn584JLIot3x45X6jsb:-Kn584JLIot3x45X6jsc:b-stj70q) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L111)):*
> ```Go
> 	var err error
> 	for _, arg := range args {
> 		parts := strings.SplitN(arg, "=", 2)
> ```

let's not do our own flag parsing. the cli package has stuff for this

---

*[pkg/ccl/sqlccl/csv.go, line 154 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn58GqRR7khhmG-6iwg:-Kn58GqRR7khhmG-6iwh:b-3nrb6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L154)):*
> ```Go
> 		return err
> 	}
> 	log.Infof(ctx, "CSV rows read: %d", csv)
> ```

I think our cli output uses fmt.Printf

---

*[pkg/ccl/sqlccl/csv.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn58nk3a4c2XzWT7Afx:-Kn58dWH3shUHDvqn1Zv:b-gv2sia) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L203)):*
> ```Go
> 
> 	const (
> 		parentID = 50
> ```

comment on these ids. also, you probably want them to be in terms of keys.MaxReservedDescID

---

*[pkg/ccl/sqlccl/csv.go, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn58xlbGNO0MbR6W2Mh:-Kn58xlc7YJiiWdKVbzt:b1ru625) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L206)):*
> ```Go
> 		id       = 51
> 	)
> 	tableDesc, err := sql.MakeTableDesc(ctx, nil /* txn */, sql.NilVirtualTabler, nil /* SearchPath */, create, parentID, id, sqlbase.NewDefaultPrivilegeDescriptor(), nil /* affected */, "" /* sessionDB */, nil /* EvalContext */)
> ```

wrap plz

---

*[pkg/ccl/sqlccl/csv.go, line 216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn596OxgYbNxIVpk8ci:-Kn596OxgYbNxIVpk8cj:bap1rx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L216)):*
> ```Go
> 	}
> 	defer func() {
> 		_ = os.RemoveAll(rocksdbDest)
> ```

log if this errors. the user probably wants to know if we left a bunch of temp files sitting around

---

*[pkg/ccl/sqlccl/csv.go, line 224 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn59KUIfob-YTw3vKyg:-Kn59KUIfob-YTw3vKyh:b-89xtrz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L224)):*
> ```Go
> 	)
> 
> 	group, ctx := errgroup.WithContext(ctx)
> ```

this pattern is really nice (errgroup + defer close chan in the func) :thumbsup: 

---

*[pkg/ccl/sqlccl/csv.go, line 224 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn59nc2FGin72p_0uRn:-Kn59nc3LMgSKMFNujwT:btc356f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L224)):*
> ```Go
> 	)
> 
> 	group, ctx := errgroup.WithContext(ctx)
> ```

overwriting the ctx here is unfortunately an antipattern. if someone adds any code that uses ctx after the .Wait call below, they'll get hard to debug errors. in the backup/restore code, i've been calling this gCtx

---

*[pkg/ccl/sqlccl/csv.go, line 256 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5A1veYcnaeRjvkJ49:-Kn5A1veYcnaeRjvkJ4A:bgqrd5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L256)):*
> ```Go
> // groupWorkers creates num worker go routines in an error group.
> func groupWorkers(ctx context.Context, num int, f func(context.Context) error) error {
> 	group, ctx := errgroup.WithContext(ctx)
> ```

does this need to be a new group? I think it's fine (and better) to use only one for the whole load

also, not sure this helper buys you anything either way. it's only used the once and seems like it's unnecessary indirection

---

*[pkg/ccl/sqlccl/csv.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5BIGkZMfX_BTpu1-A:-Kn5BIGkZMfX_BTpu1-B:b5k9chc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L270)):*
> ```Go
> func readCSV(
> 	ctx context.Context,
> 	comma, comment rune,
> ```

s/comma/separator/

and maybe s/comment/commentPrefix/

---

*[pkg/ccl/sqlccl/csv.go, line 291 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5B6bEqSQz2xjZ_AN-:-Kn5B6bEqSQz2xjZ_AN0:boy8cgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L291)):*
> ```Go
> 			defer f.Close()
> 			cr := csv.NewReader(f)
> 			if comma != 0 {
> ```

this sentinel is a little weird to me, maybe just always pass it in (and decide on comma vs other in cli parsing)

---

*[pkg/ccl/sqlccl/csv.go, line 297 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5BOWlu_0-6tJPXsIj:-Kn5BOWlu_0-6tJPXsIk:b6kmylp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L297)):*
> ```Go
> 			cr.LazyQuotes = true
> 			cr.Comment = comment
> 			for i := 1; ; i++ {
> ```

i would expect this to be 0-indexed if i saw one of these messages

---

*[pkg/ccl/sqlccl/csv.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5CL5ajQvpLRt4DYUy:-Kn5CL5ajQvpLRt4DYUz:bchhsl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L320)):*
> ```Go
> 					file: dataFile,
> 					row:  i,
> 				}:
> ```

this indentation is funky. you could construct the record before the select and just send it here to clean this up

---

*[pkg/ccl/sqlccl/csv.go, line 339 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5CZ72-LJx1fP4MsMM:-Kn5CZ72-LJx1fP4MsMN:biixyxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L339)):*
> ```Go
> }
> 
> // convertRecord converts CSV records KV pairs and sends them on kvs.
> ```

nit: on the kvCh channel.

---

*[pkg/ccl/sqlccl/csv.go, line 343 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5ChxWz5Oipxfap8o_:-Kn5ChxWz5Oipxfap8oa:b-gx2cob) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L343)):*
> ```Go
> 	ctx context.Context,
> 	records <-chan csvRecord,
> 	kvCh chan<- roachpb.KeyValue,
> ```

usage of this Ch suffix is inconsistent. let's either use it everywhere or nowhere

---

*[pkg/ccl/sqlccl/csv.go, line 352 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5DC3XMesmxD2z0SjD:-Kn5DC3XMesmxD2z0SjE:b-lz1oid) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L352)):*
> ```Go
> 	for _, col := range visibleCols {
> 		if col.DefaultExpr != nil {
> 			return errors.Errorf("column %q: DEFAULT expresion unsupported", col.Name)
> ```

spelling: expression

also maybe include the expression in the error

---

*[pkg/ccl/sqlccl/csv.go, line 374 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5DdElcKnlzwFrIM1M:-Kn5DdElcKnlzwFrIM1N:bvnh8m9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L374)):*
> ```Go
> 	parse := parser.Parser{}
> 	evalCtx := parser.EvalContext{}
> 	cols, defaultExprs, err := sqlbase.ProcessDefaultColumns(tableDesc.Columns, tableDesc, &parse, &evalCtx)
> ```

what is this for? (I actually don't know)

---

*[pkg/ccl/sqlccl/csv.go, line 380 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5Dhrr9E5UoU0RqxvL:-Kn5Dhrr9E5UoU0RqxvM:b-kv6pvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L380)):*
> ```Go
> 
> 	datums := make([]parser.Datum, len(visibleCols))
> 	var inserterr error
> ```

insertErr

---

*[pkg/ccl/sqlccl/csv.go, line 418 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5E7XSXotzv1VujFKO:-Kn5E7XSXotzv1VujFKP:b-t6as5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L418)):*
> ```Go
> 
> type sstContent struct {
> 	content []byte
> ```

I've been calling this "data"

---

*[pkg/ccl/sqlccl/csv.go, line 442 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5EmUUz7_2QxbBuqzq:-Kn5EmUUz7_2QxbBuqzr:b2nhzvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L442)):*
> ```Go
> 	b := r.NewBatch()
> 	var mk engine.MVCCKey
> 	var bs int64
> ```

bs?

---

*[pkg/ccl/sqlccl/csv.go, line 450 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5OupSLQ4ELDzQlEA1:-Kn5OupTqT9MBAels5bY:b-th323f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L450)):*
> ```Go
> 		// key. Calling .Get before each .Put is very slow. Instead, give each key
> 		// a unique timestamp. When we iterate in order, this timestamp will be set
> 		// to a static value, which will cause duplicate keys to error during sst.Add.
> ```

if you're going to rely on sst.Add to catch the duplicates, you've got to handle the case where you get one, close the current sst, start a new sst, and get the other. I think it's more clear to just do this explictly

---

*[pkg/ccl/sqlccl/csv.go, line 451 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5F5TYo-5ufrm6scsq:-Kn5F5TYo-5ufrm6scsr:b-4jjwsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L451)):*
> ```Go
> 		// a unique timestamp. When we iterate in order, this timestamp will be set
> 		// to a static value, which will cause duplicate keys to error during sst.Add.
> 		mk.Timestamp.WallTime++
> ```

You could also make a hlc.Clock and called Now for each kv instead

---

*[pkg/ccl/sqlccl/csv.go, line 457 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5Fguc08E_dgGZACuH:-Kn5Fguc08E_dgGZACuI:baggcr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L457)):*
> ```Go
> 		}
> 		bs++
> 		if bs > batchCount {
> ```

The tunings are much more obvious if you batch based on size instead of kv count. len(batch.Repr()) would do it

---

*[pkg/ccl/sqlccl/csv.go, line 479 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5PZU-XamYBFr7p3td:-Kn5PZU-XamYBFr7p3te:b-v1eqzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L479)):*
> ```Go
> 
> 	var kv engine.MVCCKeyValue
> 	var firstKey roachpb.Key
> ```

the semantics of resetting kv and firstKey are a little spread out. This method should probably just take start and end key as arguments and then they can be entirely the responsibility of the iter loop below

---

*[pkg/ccl/sqlccl/csv.go, line 495 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5GkIBNlL4u6eD8qfn:-Kn5GkIC-wQH8LAD2op9:bsqf77m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L495)):*
> ```Go
> 			span: roachpb.Span{
> 				Key:    firstKey,
> 				EndKey: kv.Key.Key.PrefixEnd(),
> ```

.Next does the same thing you're trying to accomplish but more precisely

---

*[pkg/ccl/sqlccl/csv.go, line 541 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5Q7Eu7XX5A1xk3b6b:-Kn5Q7Eu7XX5A1xk3b6c:b-xxyzyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L541)):*
> ```Go
> 		}
> 	}
> 	return count, writeSST()
> ```

writeSST needs to bail early if it doesn't have any entries (which can happen if you have exactly the wrong number of kvs so it resets right before it ends). the rocksdb code requires that there be at least 1 entry

---

*[pkg/ccl/sqlccl/csv.go, line 582 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5QcdCsyRccG0oP8PM:-Kn5QcdCsyRccG0oP8PN:bpd3z4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L582)):*
> ```Go
> 	}
> 
> 	sort.Slice(backupDesc.Files, func(i, j int) bool {
> ```

backupFileDescriptors implements sort.Interface

---

*[pkg/ccl/sqlccl/csv.go, line 596 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5R3BqNdvbE97-wjXV:-Kn5R3BqNdvbE97-wjXW:b-u8tvt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L596)):*
> ```Go
> 	backupDesc.Descriptors = []sqlbase.Descriptor{
> 		*sqlbase.WrapDescriptor(&sqlbase.DatabaseDescriptor{
> 			Name: "csv",
> ```

document this in the command line tool's help please

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-:-Kn5RAC6PTfa4fQHgLBQ:bg6gaxa)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 0 of 7 files reviewed at latest revision, 33 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kna4JKRZVqEjsUbh69Y:-Kna4JKSLm0NREC96skH:b-m5fxjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L163)):*
> ```Go
> func loadCSV(
> 	ctx context.Context,
> 	table string,
> ```

also just noticed that table should be an `io.Reader` and dataFiles should be `[]io.Reader`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 7 files reviewed at latest revision, 33 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn55h-UwFJbiupy7Nz-:-KoJ73FUzo5938XRaLRU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L43)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

This should live in a new cliccl package
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn55s4bt-5sV38pyTAf:-KoJ9p_9ZEoIZvIw7xrs:b4xxn5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L44)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I'd prefer these to all have "--" or "-" prefixes so they look more like flags. Here and below
</blockquote></details>

Was trying to keep the syntax the same as Oracle's. But I think your suggestion is better.

---

*[pkg/ccl/sqlccl/csv.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn56sAWnxFfvjXZPm4k:-KoJ8CDNmseb2pOZSxa5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L53)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Let's remove this overwrite option. It introduces tons of complexity below and doesn't add much value
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn575VrzyJRo_Li-3I2:-KoJA9VUlEPv4Mj_Ilm6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L56)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

s/temp/os temp/
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn57O2ruGeVXI6AEVeD:-KoJAEl330mtUekS5utE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L78)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

s/loading/bulk loading/ (also you can remove "Various")
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn584JLIot3x45X6jsb:-KoOEX1dsGFMQlv5BZ4z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L111)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

let's not do our own flag parsing. the cli package has stuff for this
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 154 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn58GqRR7khhmG-6iwg:-KoOEc8yP2OQI6EU-fTp:b-bx0eim) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L154)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I think our cli output uses fmt.Printf
</blockquote></details>

The thought here was that this stuff would only be printed if the user requested the logs. Most of it isn't useful to the user. Do they really care how many kv pairs and sst files were made?

---

*[pkg/ccl/sqlccl/csv.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kna4JKRZVqEjsUbh69Y:-KoOFoDbBbTipgW-Y7_1:behlc10) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L163)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

also just noticed that table should be an `io.Reader` and dataFiles should be `[]io.Reader`
</blockquote></details>

I'm not sure I agree with this. I like that this function can add a default dataFiles as table + `.dat`, which would have to be moved up. Also, I like that readCSV only has one data file open at a time. If we pass `[]io.Reader`, then all those file descriptors would have to be open the whole time, even though they are only being read one at a time.

---

*[pkg/ccl/sqlccl/csv.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn58nk3a4c2XzWT7Afx:-KoOGj2RdMMKSrAH6_wK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L203)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

comment on these ids. also, you probably want them to be in terms of keys.MaxReservedDescID
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn58xlbGNO0MbR6W2Mh:-KoOGyCpmDKT9ENoK2UU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L206)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

wrap plz
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn596OxgYbNxIVpk8ci:-KoOJtPzwRDi-1nqQF7U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L216)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

log if this errors. the user probably wants to know if we left a bunch of temp files sitting around
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 224 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn59nc2FGin72p_0uRn:-KoOK629dr2iG_IKb0ur:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L224)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

overwriting the ctx here is unfortunately an antipattern. if someone adds any code that uses ctx after the .Wait call below, they'll get hard to debug errors. in the backup/restore code, i've been calling this gCtx
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 256 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5A1veYcnaeRjvkJ49:-KoOK9MvWVfxt_-pdjqt:bsjqsk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L256)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

does this need to be a new group? I think it's fine (and better) to use only one for the whole load

also, not sure this helper buys you anything either way. it's only used the once and seems like it's unnecessary indirection
</blockquote></details>

Yes, it needs to be a new group. It needs to close its chan after all incoming rows have been processed, which means it needs to wait until all of its go routines have finished. It uses a helper here because I previously had more functions that called the helper, but it's now reduced to just this one instance. I think having the helper is more clear, and will allow us to use it more in the future when doing things like converting postgres COPY or INSERT data directly into backup files.

---

*[pkg/ccl/sqlccl/csv.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5BIGkZMfX_BTpu1-A:-KoOKuE9JlkuJpBBdUob:b-2f8k1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L270)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

s/comma/separator/

and maybe s/comment/commentPrefix/
</blockquote></details>

They are named this because those are their names in the csv package. I think it's more clear to keep the names as is since they are used exactly for that package and nothing else.

---

*[pkg/ccl/sqlccl/csv.go, line 291 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5B6bEqSQz2xjZ_AN-:-KoOL8BajqzAMfR0UxYO:b2km3lj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L291)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this sentinel is a little weird to me, maybe just always pass it in (and decide on comma vs other in cli parsing)
</blockquote></details>

The csv package defaults to comma, but I agree this is a bit weird. Will move this up.

---

*[pkg/ccl/sqlccl/csv.go, line 297 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5BOWlu_0-6tJPXsIj:-KoOLe4ntPQOieaumZcs:b-69dq59) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L297)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

i would expect this to be 0-indexed if i saw one of these messages
</blockquote></details>

It is almost the same as a line number, which are 1-indexed.

---

*[pkg/ccl/sqlccl/csv.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5CL5ajQvpLRt4DYUy:-KoOLZQBshBq-_KYdC89:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L320)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this indentation is funky. you could construct the record before the select and just send it here to clean this up
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 339 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5CZ72-LJx1fP4MsMM:-KoOLq6KubpTUPEAgOMQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L339)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: on the kvCh channel.
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 343 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5ChxWz5Oipxfap8o_:-KoOMMq3vZt-Q2M5pdAl:b-mksbsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L343)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

usage of this Ch suffix is inconsistent. let's either use it everywhere or nowhere
</blockquote></details>

Used ch everywhere.

---

*[pkg/ccl/sqlccl/csv.go, line 352 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5DC3XMesmxD2z0SjD:-KoOO0doaH4ek_IVcEsO:b7hepby) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L352)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

spelling: expression

also maybe include the expression in the error
</blockquote></details>

I'm worried that including the expression will make people think that specific expression isn't supported, when the real problem here is that no default expression is supported. (Also, this requirement can go away, so this error message will go away soon too.)

---

*[pkg/ccl/sqlccl/csv.go, line 374 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5DdElcKnlzwFrIM1M:-KoOOgPGgkhk8Zn4xQ_h:bq3rqwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L374)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

what is this for? (I actually don't know)
</blockquote></details>

We don't currently support default expressions on visible columns, but we do on hidden columns (the default _rowid one). This line sets up those expressions to run.

---

*[pkg/ccl/sqlccl/csv.go, line 380 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5Dhrr9E5UoU0RqxvL:-KoOPqrFu438Y3-8RvxR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L380)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

insertErr
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 418 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5E7XSXotzv1VujFKO:-KoOPzj6nqHEHCOs7dO2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L418)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I've been calling this "data"
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 442 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5EmUUz7_2QxbBuqzq:-KoOQORg2s0X4F6CPge-:b34uopr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L442)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

bs?
</blockquote></details>

This has been removed.

---

*[pkg/ccl/sqlccl/csv.go, line 450 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5OupSLQ4ELDzQlEA1:-KoOcu4fbAIq30EjeEKF:b-1xdbqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L450)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

if you're going to rely on sst.Add to catch the duplicates, you've got to handle the case where you get one, close the current sst, start a new sst, and get the other. I think it's more clear to just do this explictly
</blockquote></details>

I've added the between SST checking, but the timestamp stuff still needs to be here.

---

*[pkg/ccl/sqlccl/csv.go, line 451 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5F5TYo-5ufrm6scsq:-KoOV7xnw4JbP0WEFbyf:b-gmxshf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L451)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

You could also make a hlc.Clock and called Now for each kv instead
</blockquote></details>

I don't see the benefit of that because the clocks all get set to the same value later to detect duplicates.

---

*[pkg/ccl/sqlccl/csv.go, line 457 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5Fguc08E_dgGZACuH:-KoOUxW1pXh1qEBLrW40:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L457)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

The tunings are much more obvious if you batch based on size instead of kv count. len(batch.Repr()) would do it
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 479 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5PZU-XamYBFr7p3td:-KoQau-oHS1DxO2_aPit:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L479)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

the semantics of resetting kv and firstKey are a little spread out. This method should probably just take start and end key as arguments and then they can be entirely the responsibility of the iter loop below
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 495 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5GkIBNlL4u6eD8qfn:-KoOQjIvE4VUu1JkVqCv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L495)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

.Next does the same thing you're trying to accomplish but more precisely
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 541 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5Q7Eu7XX5A1xk3b6b:-KoOSBETk9z6Hs303r8O:b-llaf1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L541)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

writeSST needs to bail early if it doesn't have any entries (which can happen if you have exactly the wrong number of kvs so it resets right before it ends). the rocksdb code requires that there be at least 1 entry
</blockquote></details>

I think it already did this, but I've updated things due to another comment.

---

*[pkg/ccl/sqlccl/csv.go, line 582 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5QcdCsyRccG0oP8PM:-KoOR973l7SMZ6wKBGtW:b-38u1tx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L582)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

backupFileDescriptors implements sort.Interface
</blockquote></details>

Done. Had to export  some stuff.

---

*[pkg/ccl/sqlccl/csv.go, line 596 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5R3BqNdvbE97-wjXV:-KoOS-kwCQl9W3AjDknN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L596)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

document this in the command line tool's help please
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
ccl/sqlccl: add a load csv command

This adds the ability to convert a CSV file to enterprise BACKUP
format using only the CLI (without connecting to any server). This
is a good-enough solution for most of our current needs. A more
complicated, distributed CSV processing and sorting algorithm could
be used in the future for very large (> many hundred GB) files.

It works by creating a pipeline of channels with worker go routines
at each stage. The stages are: read the CSV into records per line,
convert the records into KV pairs, sort all KV pairs and split into
SST files, write SST files and create the backup descriptor. Only
the second stage (conversion) can be parallelized with multpile go
routines because all others require reading or writing in order.

Although we would like CSV importing (including the RESTORE step)
to not require an enterprise license, for now this code is CCL. When
we figure out exactly how we want to do that we may move it to Apache.

Closes #15933

danhhz commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 0 of 9 files reviewed at latest revision, 38 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/cliccl/load.go, line 65 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-KoiP5sDVmOxyLJrmBPP:-KoiP5sDVmOxyLJrmBPQ:b-241ick) ([raw file](https://github.com/cockroachdb/cockroach/blob/be00d9d8878450c2131680d2b4cbd33c25049a11/pkg/ccl/cliccl/load.go#L65)):*
> ```Go
> 	flags.StringVar(&csvDest, "dest", "", "destination directory for backup files")
> 	flags.StringVar(&csvNullIf, "nullif", "", "if specified, the value of NULL; can specify the empty string")
> 	flags.StringVar(&csvComma, "comma", "", "if specified, the CSV delimiter instead of a comma")
> ```

so I agree with your "lets match the csv package's naming" logic below, but `--comma=|` seems a bit much. how about `--delimiter`?

---

*[pkg/ccl/cliccl/load.go, line 94 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-KoiPSss3FQoTO4ZYK2q:-KoiPSss3FQoTO4ZYK2r:b-3y30nd) ([raw file](https://github.com/cockroachdb/cockroach/blob/be00d9d8878450c2131680d2b4cbd33c25049a11/pkg/ccl/cliccl/load.go#L94)):*
> ```Go
> 	getRune := func(s string) (rune, error) {
> 		if s == "" {
> 			return 0, nil
> ```

I find it non-obvious that we're relying on the csv packages handling of 0 values all the way out here. Lets hardcode the csv package defaults for comma and comment as the flag defaults above (or at the very least leave some kind of comment here)

---

*[pkg/ccl/sqlccl/csv.go, line 154 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn58GqRR7khhmG-6iwg:-KoiOpvm3ubiJ8tuIuR4:b-qip88k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L154)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

The thought here was that this stuff would only be printed if the user requested the logs. Most of it isn't useful to the user. Do they really care how many kv pairs and sst files were made?
</blockquote></details>

gotcha. either way is fine by me

---

*[pkg/ccl/sqlccl/csv.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kna4JKRZVqEjsUbh69Y:-KoiMO8zOh5KXxf8Yx_2:b-s3hgpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L163)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I'm not sure I agree with this. I like that this function can add a default dataFiles as table + `.dat`, which would have to be moved up. Also, I like that readCSV only has one data file open at a time. If we pass `[]io.Reader`, then all those file descriptors would have to be open the whole time, even though they are only being read one at a time.
</blockquote></details>

I was thinking ahead to when I hook this up to the sampledataccl stuff. It's fine to leave it like this and I'll revisit

---

*[pkg/ccl/sqlccl/csv.go, line 256 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5A1veYcnaeRjvkJ49:-KoiP-JVPoUVUa1ndZYd:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L256)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Yes, it needs to be a new group. It needs to close its chan after all incoming rows have been processed, which means it needs to wait until all of its go routines have finished. It uses a helper here because I previously had more functions that called the helper, but it's now reduced to just this one instance. I think having the helper is more clear, and will allow us to use it more in the future when doing things like converting postgres COPY or INSERT data directly into backup files.
</blockquote></details>

:thumbsup: 

---

*[pkg/ccl/sqlccl/csv.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5BIGkZMfX_BTpu1-A:-KoiP3Q42Re8rZJNz1h3:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L270)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

They are named this because those are their names in the csv package. I think it's more clear to keep the names as is since they are used exactly for that package and nothing else.
</blockquote></details>

:thumbsup: 

---

*[pkg/ccl/sqlccl/csv.go, line 297 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5BOWlu_0-6tJPXsIj:-KoiPxMxpJ92bwMdbA5B:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L297)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

It is almost the same as a line number, which are 1-indexed.
</blockquote></details>

:thumbsup: 

---

*[pkg/ccl/sqlccl/csv.go, line 352 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5DC3XMesmxD2z0SjD:-KoiQ4VAAfm2nbknyghX:b-pqov82) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L352)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I'm worried that including the expression will make people think that specific expression isn't supported, when the real problem here is that no default expression is supported. (Also, this requirement can go away, so this error message will go away soon too.)
</blockquote></details>

makes sense

---

*[pkg/ccl/sqlccl/csv.go, line 374 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5DdElcKnlzwFrIM1M:-KoiNO5vtBHRZ1j7hx4z:b-q3489o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L374)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

We don't currently support default expressions on visible columns, but we do on hidden columns (the default _rowid one). This line sets up those expressions to run.
</blockquote></details>

leave this as a comment, plz

---

*[pkg/ccl/sqlccl/csv.go, line 450 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5OupSLQ4ELDzQlEA1:-KoiQ8S8oqjk8jR844XA:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L450)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I've added the between SST checking, but the timestamp stuff still needs to be here.
</blockquote></details>

:thumbsup: 

---

*[pkg/ccl/sqlccl/csv.go, line 451 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5F5TYo-5ufrm6scsq:-KoiQC7Cx6SeN09h9Qw7:b-ijir5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L451)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I don't see the benefit of that because the clocks all get set to the same value later to detect duplicates.
</blockquote></details>

I was thinking hlc.Clock would work once this is distributed but I guess that's actually not the case. this is fine too

---

*[pkg/ccl/sqlccl/csv.go, line 408 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-KoiNn4WaPuwokFXwt4b:-KoiNn4WaPuwokFXwt4c:b-b9vf97) ([raw file](https://github.com/cockroachdb/cockroach/blob/be00d9d8878450c2131680d2b4cbd33c25049a11/pkg/ccl/sqlccl/csv.go#L408)):*
> ```Go
> 		// Save the first key for the span.
> 		if firstKey == nil {
> 			firstKey = make([]byte, len(it.UnsafeKey().Key))
> ```

the usual pattern for this is `firstKey = append([]byte(nil), it.UnsafeKey().Key...)`. here and below with lastKey

---

*[pkg/ccl/sqlccl/csv.go, line 439 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-KoiOLFrOSVDepV7USyr:-KoiOLFrOSVDepV7USys:bqz8lhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/be00d9d8878450c2131680d2b4cbd33c25049a11/pkg/ccl/sqlccl/csv.go#L439)):*
> ```Go
> 		}
> 	}
> 	if firstKey != nil {
> ```

if i understand what's going on here,`sst.DataSize > 0` would be more obvious

---

*[pkg/ccl/sqlccl/backup.go, line 244 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Koi6oLfZrJXJBMuGbGm:-Koi6oLfZrJXJBMuGbGn:b-x26ee3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5817711c46d0fffeff96c3cac8e8e3e37497562/pkg/ccl/sqlccl/backup.go#L244)):*
> ```Go
> 
> // BackupFileDescriptors is a slice of BackupDescriptor_File.
> type BackupFileDescriptors []BackupDescriptor_File
> ```

nit: `BackupFileDescriptors implements sort.Interface.` is probably more useful

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-:-KoiQdkF5S6dXqIUKdGp:b-uwkowd)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 9 files reviewed at latest revision, 38 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/cliccl/load.go, line 65 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-KoiP5sDVmOxyLJrmBPP:-KokCSoy3lOXzqhnqI93:b-t77xc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/be00d9d8878450c2131680d2b4cbd33c25049a11/pkg/ccl/cliccl/load.go#L65)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

so I agree with your "lets match the csv package's naming" logic below, but `--comma=|` seems a bit much. how about `--delimiter`?
</blockquote></details>

agreed

---

*[pkg/ccl/cliccl/load.go, line 94 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-KoiPSss3FQoTO4ZYK2q:-KokDADmPZsRfijfS4hK:b-f8sf1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/be00d9d8878450c2131680d2b4cbd33c25049a11/pkg/ccl/cliccl/load.go#L94)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I find it non-obvious that we're relying on the csv packages handling of 0 values all the way out here. Lets hardcode the csv package defaults for comma and comment as the flag defaults above (or at the very least leave some kind of comment here)
</blockquote></details>

Added a comment.

---

*[pkg/ccl/sqlccl/csv.go, line 374 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Kn5DdElcKnlzwFrIM1M:-KokEUk84gKz1EcrbVN3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad1a5ad7039a1868413697f6de24bedc0acb123b/pkg/ccl/sqlccl/csv.go#L374)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

leave this as a comment, plz
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 408 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-KoiNn4WaPuwokFXwt4b:-KokEvoK4lQuvRCFnjQ7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be00d9d8878450c2131680d2b4cbd33c25049a11/pkg/ccl/sqlccl/csv.go#L408)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

the usual pattern for this is `firstKey = append([]byte(nil), it.UnsafeKey().Key...)`. here and below with lastKey
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 439 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-KoiOLFrOSVDepV7USyr:-KokF1MAS_UfcSSxec6n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be00d9d8878450c2131680d2b4cbd33c25049a11/pkg/ccl/sqlccl/csv.go#L439)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

if i understand what's going on here,`sst.DataSize > 0` would be more obvious
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup.go, line 244 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611#-Koi6oLfZrJXJBMuGbGm:-KokF2rYk26PJh6kc1Qd:b-sltxp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5817711c46d0fffeff96c3cac8e8e3e37497562/pkg/ccl/sqlccl/backup.go#L244)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: `BackupFileDescriptors implements sort.Interface.` is probably more useful
</blockquote></details>

These comments all got reverted so no need to document -_-

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Reviewed 2 of 8 files at r1, 4 of 11 files at r2, 2 of 8 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 38 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16611)*
<!-- Sent from Reviewable.io -->
